### ISSUE TITLE
Error that is always caught still counts towards return ErrorSet for checking, but only sometimes

### ISSUE THREAD

**Issue by tgschultz** [ziglang/zig#1810]
Edit: see my comments later on. It is fair that I should have to @errSetCast in this instance, however the fact that I don't have to under some circumstances made this problem slip through testing.

```
pub fn testErrSet0(x: u8) error{ErrorA,ErrorB}!void {
    if (x > 1) return error.ErrorA;
     return error.ErrorB;
}

pub fn testErrSet1(x: u8) error{ErrorA}!void {
    testErrSet0(x) catch |err|
    {
        if (err == error.ErrorB) return;
        return err;
    };
}

test "" {
    try testErrSet1(0);
}
```
```
test.zig:23:16: error: expected type '(anonymous error set at test.zig:18:27)', found '(anonymous error set at test.zig:12:27)'
        return err;
               ^
test.zig:12:40: note: 'error.ErrorB' not a member of destination error set
pub fn testErrSet0(x: u8) error{ErrorA,ErrorB}!void
                                       ^
```

**Comment by Hejsil** [comment]
Related #769, though this is broader in scope and would probably require control flow analysis.

**Comment by tgschultz** [comment]
I think it's actually fair that I should have to cast it, but was unaware of @errSetCast.

What's weird about this is that it came up when working with code from #1775, but was never caught in any of the tests. The only explanation I can think of for this is that I used SliceInStream for the tests and I'm using a File stream in the code that triggered this. File stream has a sizeable errorset, while SliceInStream's errorset is empty. `error.EndOfStream` is returned by `InStream(Error).readNoEof` which is what ultimately triggers this issue. `readNoEof` has an inferred errorset. It's as though the empty errorset somehow causes incorrect inferrence up the chain of inferred errorsets.

**Comment by Hejsil** [comment]
@tgschultz Yea, streams and error sets are a pain right now. You also can't use things like a `NullOutStream` if `stderr` isn't available, as their streams are of different types.

**Comment by andrewrk** [comment]
Here's @Sahnvour's proposal to fix this: https://github.com/ziglang/zig/issues/1923#issuecomment-461530246

This is what is planned

**Comment by andrewrk** [comment]
I don't know what I was thinking with https://github.com/ziglang/zig/issues/1810#issuecomment-467655104. This issue is not a bug nor is it related to error return traces.

**Comment by andrewrk** [comment]
I believe this is solved by #769.

When zig has advanced control flow analysis, then it should be taken advantage of to omit some errors from inferred error sets.
